closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -145,22 +161,26 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_DigestUpdate
 
     OpenSSLMDContext *context = (OpenSSLMDContext*) c;
 
+    if (context == NULL)
+        return -1;
+
     if (message == NULL) {
-        // Data passed in through direct byte buffer
-        if (1 != EVP_DigestUpdate(context->ctx, context->nativeBuffer, messageLen))
-            handleErrors();
-    } else {
-        unsigned char* messageNative = (*env)->GetPrimitiveArrayCritical(env, message, 0);
-        if (messageNative == NULL) {
-            return -1;
-        }
+        return -1;
+    }
 
-        if (1 != EVP_DigestUpdate(context->ctx, (messageNative + messageOffset), messageLen))
-            handleErrors();
+    unsigned char* messageNative = (*env)->GetPrimitiveArrayCritical(env, message, 0);
+    if (messageNative == NULL) {
+        return -1;
+    }
 
+    if (1 != EVP_DigestUpdate(context->ctx, (messageNative + messageOffset), messageLen)) {
+        printErrors();
         (*env)->ReleasePrimitiveArrayCritical(env, message, messageNative, 0);
+        return -1;
     }
 
+    (*env)->ReleasePrimitiveArrayCritical(env, message, messageNative, 0);
```

Comment:
The uses of `ReleasePrimitiveArrayCritical` should use `JNI_ABORT` because the array is only read and there's no need for the VM to copy data back to the array in the case a copy was made by `GetPrimitiveArrayCritical`.